modules/imgcodecs/src/grfmt_png.cpp

```diff
@@ -449,6 +941,824 @@ bool  PngEncoder::write( const Mat& img, const std::vector<int>& params )
+    {
+        uchar* pa = pimage1;
+        uchar* pb = pimage2;
+        uchar* pc = ptemp;
+
+        for (j = 0; j < h; j++)
+            for (i = 0; i < w; i++)
+            {
+                uint c1 = (pa[2] << 16) + (pa[1] << 8) + pa[0];
+                uint c2 = (pb[2] << 16) + (pb[1] << 8) + pb[0];
+                if (c1 != c2)
+                {
+                    diffnum++;
+                    if (has_tcolor && c2 == tcolor)
+                        over_is_possible = 0;
+                    if (i < x_min)
+                        x_min = i;
+                    if (i > x_max)
+                        x_max = i;
+                    if (j < y_min)
+                        y_min = j;
+                    if (j > y_max)
+                        y_max = j;
+                }
+                else
+                    c2 = tcolor;
+
+                memcpy(pc, &c2, 3);
+                pa += 3;
+                pb += 3;
+                pc += 3;
+            }
+    }
+    else if (bpp == 4)
+    {
+        uint* pa = (uint*)pimage1;
+        uint* pb = (uint*)pimage2;
+        uint* pc = (uint*)ptemp;
+
+        for (j = 0; j < h; j++)
+            for (i = 0; i < w; i++)
+            {
+                uint c1 = *pa++;
+                uint c2 = *pb++;
+                if ((c1 != c2) && ((c1 >> 24) || (c2 >> 24)))
+                {
+                    diffnum++;
+                    if ((c2 >> 24) != 0xFF)
+                        over_is_possible = 0;
+                    if (i < x_min)
+                        x_min = i;
+                    if (i > x_max)
+                        x_max = i;
+                    if (j < y_min)
+                        y_min = j;
+                    if (j > y_max)
+                        y_max = j;
+                }
+                else
+                    c2 = 0;
+
+                *pc++ = c2;
+            }
+    }
+
+    if (diffnum == 0)
+    {
+        x0 = y0 = 0;
+        w0 = h0 = 1;
+    }
+    else
+    {
+        x0 = x_min;
+        y0 = y_min;
+        w0 = x_max - x_min + 1;
+        h0 = y_max - y_min + 1;
+    }
+
+    deflate_rect_op(pimage2, x0, y0, w0, h0, bpp, stride, zbuf_size, n * 2);
+
+    if (over_is_possible)
+        deflate_rect_op(ptemp, x0, y0, w0, h0, bpp, stride, zbuf_size, n * 2 + 1);
+}
+
+void PngEncoder::deflate_rect_fin(int deflate_method, int iter, uchar* zbuf, uint* zsize, int bpp, int stride, uchar* rows, int zbuf_size, int n)
+{
+    uchar* row = op[n].p + op[n].y * stride + op[n].x * bpp;
+    int rowbytes = op[n].w * bpp;
+
+    if (op[n].filters == 0)
+    {
+        uchar* dp = rows;
+        for (int j = 0; j < op[n].h; j++)
+        {
+            *dp++ = 0;
+            memcpy(dp, row, rowbytes);
+            dp += rowbytes;
+            row += stride;
+        }
+    }
+    else
+        process_rect(row, rowbytes, bpp, stride, op[n].h, rows);
+
+    if (deflate_method == 2)
+    {
+        CV_UNUSED(iter);
+        /*ZopfliOptions opt_zopfli;
```

Comment:
If you have a commented block, put it in an `#If 0` and please explain why it is now commented out.